### ISSUE TITLE
PROJ version 5 vs. PROJ.4

### ISSUE THREAD

**Issue by busstoptaktik** [OSGeo/PROJ#735]
**TL;DR: Some problems associated with the PROJ.4-to-PROJ name change can be worked around by adopting a few *additional* naming conventions.**

Rationale
---------

About a month ago, @kbevers and @hobu conducted a discussion on the mailing list, ending up in adopting that the next version of PROJ.4 will not be PROJ.4 version 5.0.0, but PROJ version 5.0.0.

I didn’t take part in the discussion, as I was quite busy actually developing what is now about to become version 5.0.0, so I'm sorry for being late to the party, but I have some additional input which I think is relevant.

I also think this input first and foremost calls for developer consensus, which is why I post it as an Github issue, rather than a mailing list discussion item (obviously, the mailing list is where it will go, once I have seen developer consensus).

The Problem
-----------

The problem with the name change is that, while decoupling the name of PROJ with the version number buys us a communication advantage, it also buys us a lot of trouble which, I believe, can be worked around by adopting a few more naming conventions.

The trouble is that the name change invalidates the connection between PROJ and the URLs associated with PROJ. That is (mostly) http://proj4.org and https://github.com/OSGeo/proj.4, but also (probably) a number of OSGeo pages, and e.g. the Wikipedia page https://en.wikipedia.org/wiki/PROJ.4.

Also, both the PROJ.4 and PROJ names are, in various combinations, baked into build- and packaging systems, where it will be a hassle (and largely out of our hands) to change - cf. the mailing list discussion http://osgeo-org.1560.x6.nabble.com/Time-for-a-new-release-td5341895.html.

Solution Proposed
-----------------

The solution I propose is to **keep PROJ.4** - not as the name of the *product*, but as the name of the *project*. With a few more straightforward naming conventions, the message we will communicate is this:

* We are the PROJ.4 project
* We provide the product **PROJ**, which is now at version 5.0.0
* The foundational component of PROJ is the library `libproj`
* Other components include the application `proj`, which provide a command line interface to `libproj`
* You can contribute to the project by e.g. providing test data, writing documentation, suggesting improvements, helping fellow users on the mailing list, and by submitting code improvements and extensions

By **switching** the name of the product from PROJ.4 to PROJ, we gain the advantage of a product name that does not need an explanation to go with it.

By **keeping** PROJ.4 as the name of the project, rather than the product, we avoid a lot of unneeded trouble with web presence etc.

Keeping PROJ.4 as an identifier for the project is also a tribute in the memory of project founder Gerald I. Evenden.

Invitation
----------

So, @kbevers, @hobu, @rouault, @cffk, @aaronpuchert, @micahcochran, @QuLogic, @mloskot, @warmerdam: Could you live with being part of the PROJ.4 project, providing the PROJ product, including the `libproj` library and the `proj` application? Or do you see better solutions?

**Comment by mloskot** [comment]
@busstoptaktik
> Could you live with being part of the PROJ.4 project, providing the PROJ product, including the libproj library and the proj application?

Yes.

**Comment by aaronpuchert** [comment]
Since we already have `libproj` and `proj` as filenames, I don't see any compatibility problems arising from the decision. The package has also always been called [proj](https://build.opensuse.org/package/show/Application:Geo/proj) for us. Maybe someone can find a nice [backronym](https://en.wikipedia.org/wiki/Backronym) for PROJ.4, so we have a good explanation for the project name in the future.

**Comment by cffk** [comment]
The current (version 4.x) way cmake projects utilize proj.4 is with the idiom

    find_package (PROJ4 ...)
    include_directories (${PROJ4_INCLUDE_DIRS})
    target_link_libraries (... ${PROJ4_LIBRARIES})

The original proposal by @kbevers and @hobu would have meant dropping the "4" here.  I was gearing up to propose some transition period where both the PROJ4 and PROJ names would have been supported.

However I believe that @busstoptaktik is suggesting that, in this context, the PROJ4 name would be retained (with no need to transition to the PROJ name).  I am in favor of this.

P.S., I have no problem with an idiom

    find_package (PROJ4 5.0)

**Comment by hobu** [comment]
> P.S., I have no problem with an idiom
> ```find_package (PROJ4 5.0)```

Me either.

**Comment by kbevers** [comment]
I am good with the proposed solution.

**Comment by mloskot** [comment]
Recent thread related to this issue, [New OSGeo website should rename Proj4 to Proj](http://lists.maptools.org/pipermail/proj/2018-January/007938.html) calls to:
> Please rename the project on the new website from "Proj4" to "Proj".

**Comment by sebastic** [comment]
> Keeping PROJ.4 as an identifier for the project is also a tribute in the memory of project founder Gerald I. Evenden.

IIRC he founded the project PROJ, which had four major releases known as PROJ, PROJ.2, PROJ.3 & PROJ.4.

I think that the project should also be renamed from PROJ.4 to PROJ. The marjor version name has no place in the project name nor product name.

The web presences can easily be renamed in most cases. Over time most references to PROJ.4 will be replaced by PROJ.

**Comment by cffk** [comment]
According to Wikipedia, it's been proj.4 since 1994 which means that version 4 is the _only_ version that most programmers have ever encountered.  Since then the _branding_ has invariably included the "4".  So I recommend keeping this convention while allowing the version to increment beyond 4.

**Comment by sebastic** [comment]
In [Cartographic Projection Procedures for the UNIX Environment — A User’s Manual](http://download.osgeo.org/proj/OF90-284.pdf) Gerald refers to "the cartographic projection program proj (release  3)" aka PROJ.3.

Since the project was named after the program, it should just be "proj" and not include the major version.

**Comment by kbevers** [comment]
> Since the project was named after the program, it should just be "proj" and not include the major version.

Ultimately it is up to us, as a community, to decide what the name should or shouldn't be. If a consistent name is what is preferred by the majority, that is what we should use. And when version 5 is released the major version number will no longer be a part of the name.

I personally value the consistency of the name higher than a logical name (which it isn't anyway, since we do much more than projections these days).

**Comment by aaronpuchert** [comment]
I've been thinking about this again, and it occurred to me that people who are just getting started with GIS in general or proj in particular are going to be quite confused by a program/library/project variably called proj, PROJ, proj4, proj.4 or PROJ.4, which is coming in version 5.0.

There are two groups of people to consider here: those that have used proj for many years, and those who have just started using it or are going to use it in the future. The former group wants as little change as possible, while the latter group would probably welcome a consistent experience with as few surprises as possible. Of course the people commenting in this thread are heavily biased towards the former group, so I wanted to voice the concerns of possible future users, who might consider this a mess and have little understanding why one would insist on some arguably outdated name.

Right now it seems like the easier way to keep the PROJ.4, or proj.4 project name, but a few years down the road this will feel like historical baggage in need of constant explanation, because there are always new people coming in failing to grasp why the project still has an old version number in the name although it has already moved on.

> Since then [1994] the branding has invariably included the "4".

One could counter that the number was always understood as a version number, and people are aware that version numbers change. Most Linux distributions have wisely kept it out of the package name: [Ubuntu](https://packages.ubuntu.com/artful/proj-bin), [Debian](https://packages.debian.org/wheezy/proj-bin), [OpenSUSE](https://build.opensuse.org/package/show/Application:Geo/proj), [Fedora](https://admin.fedoraproject.org/pkgdb/package/rpms/proj/). And that number was never part of either the program's or the library's name.

**Comment by busstoptaktik** [comment]
> IIRC he founded the project PROJ, which had four major releases known as PROJ, PROJ.2, PROJ.3 & PROJ.4.

@sebastic: I don't think there ever was a PROJ *project* intended from Gerald before he retired from USGS: The **proj** Unix style filter program, I believe, was intended as a helper application for his **MAPGEN** map plotting software, and probably started out as part of that project.

I have never seen **proj** releases prior to 3, and I do not think those releases were ever "releases" in the sense of being circulated outside of Gerald's department of the USGS.

I used **proj** release 3 in the early 1990s, but did not use the related library until the spring of 1995, where release 4 was already widely circulated.

**After** Gerald retired, he started the project `libproj4`. Its documentation is part of the PROJ.4 distribution - it can be found in `docs/old/libproj.pdf`. And despite the name of the pdf, the title of the work is **libproj4: A Comprehensive Library of Cartographic Projection Functions**

Also, in the spring of 2008 I had a long e-mail conversation with Gerald, where he talks about his project as `libproj4`. I talked to Gerald at the suggestion of  Frank @warmerdam as I was trying to get Frank to accept the *Engsager Extended Transverse Mercator* code (now in `proj_etmerc.c`) into Frank's PROJ.4 distribution.

Frank preferred to track the projection material from Gerald's project, while keeping his own extensions (3D, datums) as an add-on. Hence, the way to get the code into `PROJ.4` was to go to Gerald, and get him to accept it into his project, `libproj4`.

Eventually, Gerald lost interest in maintaining the material, but the `libproj4` extensions to the original *PROJ Cartographic Projection Software, Release 4* now survive as part of PROJ.4, and can be found in `proj_etmerc.c`, `proj_mdist.c`, `proj_rouss.c`, `PJ_omerc.c`, and perhaps a few more.

> I think that the project should also be renamed from PROJ.4 to PROJ. The marjor version name has no place in the project name nor product name.

I think differently, and I do not necessarily interpret the digit 4 in the name as an indication of a major version number.

I would, on the other hand, be very willing to see the digit as a nod to Gerald, and his `libproj` project, as well as being consistent with the *de facto* brand name for a group of people that has maintained the heritage from Gerald for many years.

> The web presences can easily be renamed in most cases. Over time most references to PROJ.4 will be replaced by PROJ.

I take the "easily" as an indication that you volunteer to do it.

But since proj.org, proj.net, proj.com, and a number of other relevant domains are all owned by people with no relation to PROJ, renaming proj4.org to proj.org is not an option, which means losing a primary web presence if calling the project PROJ.

On Github, I think backing up an entire repo, including issues, pull requests, inline discussions, etcetera, can be done, so I'm not quite as concerned wrt. the Github presence...

I would just not have any idea where to start migrating from http://github.com/OSGeo/proj.4 to http://github.com/OSGeo/proj

(and really would prefer spending time writing documentation, improving, and extending code, rather than wasting it on doing a migration that would mean losing the primary web presence).

> In Cartographic Projection Procedures for the UNIX Environment — A User’s Manual Gerald refers to "the cartographic projection program proj (release 3)" aka PROJ.3.

As I have indicated above, this is not the only relevant naming element. Gerald's last project, and the one Frank indicated to track, was libproj4.

> Since the project was named after the program, it should just be "proj" and not include the major version.

I do not think there ever *was* a PROJ.4 **project**: This is actually what I suggest to start now. Organizationally, PROJ.4 is a part of the MetaCRS project which, after umpteen years has still not graduated from OSGeo incubation.

As PROJ.4 is the geodetic/cartographic foundation of most other OSGeo projects, I find this situation absurd, and think another (perhaps supplementary) organizational framework than the apparently dormant MetaCRS would be a nice way forward.

So, in brief, @sebastic, I think I have refuted your arguments above. You may disagree, and in that case I would be honored if you would agree to disagree with me.

**Comment by pl-nfwarmerdam** [comment]
For what it's worth I contemplated the problem of having a major version as part of the project name a number of times over the years, and came to the conclusion that the library/project is effectively known as PROJ.4, and we should just avoid ever bumping the major version beyond 4, and live with the peculiarity.  "PROJ" on it's own is so generic as to be not meaningful.  Of course, I have little authority on this, and the whole matter is very subjective despite the significant efforts to construct rationales for various approaches.

**Comment by busstoptaktik** [comment]
@pl-nfwarmerdam / @warmerdam: nice hearing from you, and nice to hear that we agree that PROJ.4 is the obvious brand name, and that trying to establish a different brand (PROJ) will be hard (for various reasons, including the genericity of the name).

There is no reason to construct rationales for keeping the PROJ.4 moniker around (as the project name or otherwise): I do not agree with @sebastic that switching the web presence will be *easy*. We need to keep a connection to the branding that is there already.

My reason for starting this thread is that I want to avoid having to waste huge amounts of time rebranding, now that it *has* been decided to bump the major version number to 5. It is more a matter of damage control than anything else.

I'm not particularly in favor of going to version 5, but I am even more against dropping the well established brand name PROJ.4 - that will lose us much (including our primary web presence) and bring us little good.

So my main point is that we will have to keep a connection to PROJ.4 - either by claiming that "it's the name of the project", or by any other means anyone can come up with.

I think the two line description that *"PROJ.4 is the project, PROJ is the product, it consists of the library libproj, and a number of command line tools, including proj"* is, as far I can see,  the least unclear and most brief way to make a meaningful communication effort out of this rather messy situation.

proj4.org will continue to be our primary web presence, since proj.org is not available. When people ask
> "Why is the website called proj4.org, when the software is at version 5?", 

rather than telling them 
> "You don't have to worry"

(which is true, but will make anyone worry), we can have that brief and reasonably meaningful description (much like the backronym, @aaronpuchert suggested).

The alternative would be to tell the entire story, starting in 1983, and losing people before reading to the end of it.

So keep PROJ.4 as part of the story - by making it the project name, or by any other means anyone can come up with. Dropping it entirely is just not worth the effort.

**Comment by mloskot** [comment]
I can't resist a tiny brain dump, please ignore if irrelevant:

* Drop the `dot` from PROJ.4.
* Make it read PROJ4.
* Read it backwards as the library is "For Projecting" since, obviously, 4PROJ is weird and troublesome as C/C++ software related name.
* Release PROJ4 v5.0.0
* Keep the Community less/un-confused.

PROJ4 allows consistent naming for Unix library `libproj4` and Windows `proj4.dll`.

Unfortunately, Linux package maintainers have fostered `proj` and `libproj` so that would make them greatly unhappy, I guess.

UPDATE: The Linux package maintainers have not been consistent themselves. In Ubuntu, for example. [proj4 in package names is also used](https://packages.ubuntu.com/search?keywords=proj4).

**Comment by kbevers** [comment]
> For what it's worth I contemplated the problem of having a major version as part of the project name a number of times over the years, and came to the conclusion that the library/project is effectively known as PROJ.4, and we should just avoid ever bumping the major version beyond 4, and live with the peculiarity. "PROJ" on it's own is so generic as to be not meaningful. 

When we first started this discussion months ago, I suggested keeping the name, and continuing the minor version number as the major version number, such that the next release would be "PROJ.4 v. 10.0.0". I still feel that is a good solution that at the same time honours the history and gives us a somewhat clean slate. Although at the expense of a wierd Windows-esque version number bump. My suggestion didn't really gain any traction at the time and probably won't now. 

> Of course, I have little authority on this,

I say quite the contrary. You have probably put more hours in this project than the rest of us combined and have more than earned the right to be listened to in this matter!

**Comment by busstoptaktik** [comment]
@mloskot, I really like your comment "Release PROJ4 v5.0.0", somewhat insisting that **PROJ4 is just a name. It does not necessarily mean anything. It's just a name, and now it's at version 5.0.0**.

Unfortunately it's too late to do that, and we have to find another way of avoiding throwing away brand identity.

Actually, agree even more with @kbevers: bumping all the way up to PROJ.4 version 10.0.0, making it not just clear, but evidently clear, that things have changed.

IMHO it makes lots of sense to put [semantic meaning](http://semver.org) into a version number, but not very much sense to insist on semantics of a *name*: A name can have lots of ethymological meaning, without one having to enforce that meaning semantically onto the bearer of the name.

For example, ethymologically speaking, the name of one of my daughters, is the Danish word for *skylark*. That does not make me complain she does not fly around in the summer sky.

Even though there is an ethymological (or historical) connection between the girl's name and the skylark, really to me *it is just a name - the name I use when I call my daughter*.

Much like, when I use the name PROJ.4, I am aware that there is a historical reason behind the name, related to the background as a projection library.

Today, however, PROJ.4 is so much more than a projection library: For more than 15 years, PROJ.4 has been handling a large subclass of general geodetic transformations, and at next release, through the `proj=pipeline` mechanism, that subclass will be extended enormously.

So for more than 15 years, it has not made much sense to read any kind of figurative/literal meaning into the name PROJ.4 - *it's just a name*. It could have been Frank, Gerald, or even Louis XIV, and we could usefully have kept the name, and bumped the version.

**Comment by kbevers** [comment]
> Unfortunately it's too late to do that, and we have to find another way of avoiding throwing away brand identity.

In what way is that too late? It is basically just keeping status quo. Today we have "PROJ.4 v4.9.3", tomorrow we'll have "PROJ.4 v5.0.0".

**Comment by busstoptaktik** [comment]
> In what way is that too late?

@kbevers: because a decision was taken by the maintainer on 2016-11-10 when @hobu announced on the mailing list that:

> This is enough consensus to satisfy me. PROJ v5.0.0 it is!

After a recent re-reading of the [thread](http://osgeo-org.1560.x6.nabble.com/Time-for-a-new-release-td5341895.html) (which I did not participate in back then until after the decision had been made and the subject had diverged), I believe Howard took the right decision: It is PROJ, rather than PROJ.4 **by popular demand**, and I agree that we will have to fulfill that demand.

What we will call the project, the work group, the Github team, or whatever we see ourselves as, is, on the other hand, much more up to ourselves.

And if we can use that freedom to minimize the loss of brand identity, loss of web presence, and amount of futile work needed, I think that is what we should do.

I think (as described numerous times in the preceeding posts) an obvious way of doing that is to let the project (or work group or whatever) take over the name that is now abandoned by the software, keep `http://proj4.org` and `https://github.com/OSGeo/proj.4` as our primary web presence...

**...and get on with the endless work of (and joy of) improving the PROJ code and documentation.**

And to cut a bit through the noise: Please give this post a thumbs up if you consider yourself a PROJ contributor and, all things considered, (mostly) agree that...

**...we are the PROJ.4 pro*ject*, we provide the PROJ pro*duct*, consisting of the library `libproj`, and a number of command line utilities, including the `proj` projection filter**

and a thumbs down if you, all things considered, (mostly) disagree.

UPDATE: yes, I thumbs-upped this myself, since I obviously agree.

**Comment by cffk** [comment]
+1

**Comment by kbevers** [comment]
From this discussion and subsequent informal vote I conclude that Thomas' proposed PROJ.4/PROJ/proj/libproj naming scheme offers "the path of least resistance", leaving most interested parties happy. I regard non-voters as being indifferent to the issue.

The perfect solution to the naming issue doesn't exists and what is described in this issue is a way forward that most of us can live with. I hope that those who prefer a different approach at least can appreciate the improved clarity of the various names used in relation to this software package. 

I will work this into the release notes somehow.
